---
title: "RNA-seq Differential Gene Expression Analyses"
author: "Abraham Quaye"
date: "`r Sys.Date()`"
output:
    pdf_document: default
    word_document: default
header-includes:
  - \usepackage{caption}
  - \captionsetup[figure]{labelformat=empty}
  - \usepackage{placeins}
    
---

```{r setup, echo=FALSE,include=FALSE, message=FALSE, warning=FALSE}
library(knitr)
library(flextable)
library(here)
library(ftExtra)
opts_chunk$set(echo = FALSE, message = FALSE, warning = FALSE)

# provide the correct path to the R script on your computer
source(here("scripts/r_code/r77q_degs.R"))
source(here("scripts/r_code/plot_heatmaps.R"))
source(here("scripts/r_code/enrichment_analyses.R"))
```

## Part 1. Overview Plots

### 1.1 Volcano Plot

```{r}
volcano <- list.files(here("results/figures"), pattern = "volcano_72hr",
           full.names = T)

```
![](`r volcano`)

### 1.2 Heatmap

```{r}
heatmap72 <- list.files(here("results/figures"), pattern = "heatmap72",
           full.names = T)

```
![](`r heatmap72`)

### 1.3 Barplot of DEGs 

```{r fig.height=6.5, fig.width=8.5}
deg_bar_plt
```

## Part 2. KEGG Pathway Analysis Results

### 2.1 Enriched Pathways from the Significant Upregulated Genes   
There was no significant pathway
```{r}
up_kegg_plt <- list.files(here("results/figures"), pattern = "_up_kegg.pdf",
           full.names = T)

```
![This results is obtained from only the upnregulated DEG list](`r up_kegg_plt`){height=80%}

```{r}
# save the table as a file
write.table(up_kegg_tab, file = here("results/tables/up_kegg_table.tsv"),
            sep = "\t", row.names = F, quote = F)

make_table <- function(tab){
    set_flextable_defaults(font.size = 4, table_align = "left")
    tab %>%
        mutate(geneID = gsub("/", ", ", geneID),
               across(RichFactor:zScore, ~round(.x, digits = 2)),
               across(pvalue:qvalue, ~formatC(.x, digits = 2, format = "E")),
               across(pvalue:qvalue, ~str_replace(.x,
                                                  "(\\d\\.\\d{2})E-(\\d{2})",
                                                  "\\1 x 10^-\\2^")),
               across(pvalue:qvalue, ~str_replace(.x, "\\^-0", "^-"))) %>% 
        flextable(.) %>%
        theme_box() %>%
        ftExtra::colformat_md(j = c("pvalue", "p.adjust", "qvalue"), part = "body") %>%
        flextable::align(align = "center", part = "all") %>% 
        flextable::width(width = 0.35) %>% 
        flextable::width(j = "geneID", width = 1.5)
}

make_table(up_kegg_tab)
```
\FloatBarrier

### 2.2 Enriched Pathways from the Significant Downregulated Genes
```{r}
down_kegg_plt <- list.files(here("results/figures"), pattern = "_down_kegg.pdf",
           full.names = T)

```
![This results is obtained from only the donwnregulated DEG list](`r down_kegg_plt`){height=80%}

```{r}
# save the table as a file
write.table(down_kegg_tab, file = here("results/tables/down_kegg_table.tsv"),
            sep = "\t", row.names = F, quote = F)

make_table(down_kegg_tab)
```
\FloatBarrier
### 2.3 Enriched Pathways from All the Significant Genes
```{r}
all_kegg_plt <- list.files(here("results/figures"), pattern = "_all_kegg.pdf",
           full.names = T)

```
![](`r all_kegg_plt`){height=80%}

```{r}
# save the table as a file
write.table(all_kegg_tab, file = here("results/tables/all_kegg_table.tsv"),
            sep = "\t", row.names = F, quote = F)

make_table(all_kegg_tab)
```
\FloatBarrier

### 2.4 KEGG Pathway enrichment Plots    
Generate and save pathway maps for all significantly enriched pathways identified from all significant DEGs (**D**ifferentially **E**xpressed **G**enes) combined together
```{r}
plts <- map(as.list(pull(all_kegg_tab, ID)), plot_kegg_pathway)
```

```{r}
kegg_figs <- tibble(figs = list.files(here("scripts/r_code"),
                         "^hsa.+\\.pathview\\.png$",
                         full.names = T),
       kegg_id = sub(pattern = "^/.+(hsa\\d+\\.pathview\\.png$)",
                     replacement = "\\1",
                     x = figs)
)
```

![File name: `r kegg_figs$kegg_id[[1]]`](`r kegg_figs$figs[[1]]`)

![File name: `r kegg_figs$kegg_id[[2]]`](`r kegg_figs$figs[[2]]`)

![File name: `r kegg_figs$kegg_id[[3]]`](`r kegg_figs$figs[[3]]`)

![File name: `r kegg_figs$kegg_id[[4]]`](`r kegg_figs$figs[[4]]`)

![File name: `r kegg_figs$kegg_id[[5]]`](`r kegg_figs$figs[[5]]`)
\FloatBarrier
\newpage

## Part 3: GO Enrichment Analysis Results   
### 3.1 Enriched GO Terms from the Significant Upregulated Genes  
No results generated

### 3.2 Enriched GO Terms from the Significant Downregulated Genes 
```{r}
down_go_plt <- list.files(here("results/figures"),
                          pattern = "_down_(bp|cc|mf).pdf",
                          full.names = T)

```
![](`r down_go_plt[[1]]`){height=90%}

![](`r down_go_plt[[2]]`){height=90%}

![](`r down_go_plt[[3]]`){height=90%}
\FloatBarrier

```{r}
# save the table as a file
write.table(down_go_all_table, file = here("results/tables/down_GO_table.tsv"),
            sep = "\t", row.names = F, quote = F)

make_table(down_go_all_table)
```

### 3.3 Enriched GO Terms from All the Significant Genes 
```{r}
all_go_plts <- list.files(here("results/figures"),
                          pattern = "_all_(bp|cc|mf).pdf",
                          full.names = T)

```
![](`r all_go_plts[[1]]`){height=90%}

![](`r all_go_plts[[2]]`){height=90%}

![](`r all_go_plts[[3]]`){height=90%}

```{r}
# save the table as a file
write.table(all_go_all_table, file = here("results/tables/all_GO_table.tsv"),
            sep = "\t", row.names = F, quote = F)

make_table(all_go_all_table)
```

### 3.4 Interconnections of GO Terms from the Combined Gene List (All DEGs)
```{r fig.height=18, fig.width=20}
# change the path to the folder where you'd like to save the figure
ggsave(plot = go_connections, here("results/figures/go_bp_connections.pdf"),
       width = 20, height = 18)
go_connections
```
